packages/nimble-components/src/table/index.ts

```diff
@@ -66,15 +103,71 @@ export class Table<
 
         // Ignore any updates that occur prior to the TanStack table being initialized.
         if (this.tableInitialized) {
-            this.updateTableOptions({ data: this.data });
+            this.trySetData(this.data);
+        }
+    }
+
+    public checkValidity(): boolean {
+        return Object.values(this.validity).every(x => x === false);
+    }
+
+    private trySetData(newData: TData[]): void {
+        const areIdsValid = this.validateRecordIds();
+        if (areIdsValid) {
+            this.updateTableOptions({ data: newData });
+        } else {
+            this.updateTableOptions({ data: [] });
         }
     }
 
+    private validateRecordIds(): boolean {
+        // Start off by assuming everything is valid.
+        this.duplicateRowId = false;
+        this.missingRowId = false;
+        this.invalidRowId = false;
+
+        if (this.idFieldName == null) {
+            return true;
+        }
+
+        const ids = new Set<string>();
+        for (const record of this.data) {
+            if (
+                !Object.prototype.hasOwnProperty.call(record, this.idFieldName)
+            ) {
+                this.missingRowId = true;
+                return false;
+            }
+
+            const id = record[this.idFieldName];
```

Comment:
If you haven't already, could you double check corner cases like `id === null` and `id === ''`? I'm guessing `null` would result in `invalidRowId` and `''` would be a valid id as long as no other record also uses it, but JS type rules and Set behavior might produce a different outcome.

Reply:
I think we'll actually run into problems with a client trying to have one of their IDs be an empty string (or null). If our `getRowId` function returns a falsey value, TanStack will generate a value for the ID. I'll add a check in the code for this along with a test.

Reply:
I had a related [comment](https://github.com/ni/nimble/pull/947#discussion_r1066447569) in @atmgrifter00's PR (i.e. Is empty string actually an invalid record name?).

Can we make sure to clarify the edge cases around empty strings as field names / id values and maybe capture it in specs and tests?

- are record field names allowed to be empty string? `table.data= [{"": 1}, {"": 2}]`
- is the idFieldName allowed to be empty string? `table.data= [{"": 1}, {"": 2}]; table.idFieldName = ''`
- is empty string a valid id for id value type string? `table.data= [{id: "a"}, {id: ""}]; table.idFieldName = 'id'`
- are ieee754 values allowed for id for type number? `table.data= [{id: 1}, {id: Infinity}, {id: -Infinity}, , {id: -0}, {id: +0}, {id: NaN}]; table.idFieldName = 'id'`
  - I think -Inf and +Inf are fine as values
  - We should not distinguish between -0 and +0, they should be treated as the same value
    - ie. `Object.is(-0, 0) // false` while `-0 === 0 // true`
  - NaN should not be allowed as a value for id
  - We should not allow id values outside the safe integer range
  - Edit: Actually maybe we should just only allow integer values as number ids? And enforce they are all safe with [isSafeInteger()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger#:~:text=Otherwise%20false%20.-,Description,%2C%20inclusive%20(%C2%B19%2C007%2C199%2C254%2C740%2C991).)
    - I can maybe see this being important if the integers are coming from the server as 64-bit int values by accident which cannot be safely represented as JS numbers.
- what types are allowed for the id?
    - I assume definitely: string and number
    - I assume not: null, undefined, boolean
    - Should date? Seems to have the characteristics to allow for it.

Reply:
As this is mostly edge cases and tests, I'll assume we won't address it in this PR but handle in a follow-up

Reply:
To answer some of the questions you've posed here:
- This PR assumes that record field names can be an empty string. Based on a previous comment in this PR, I thought that was what we wanted. I have a test to ensure the empty field name works correctly from a validation point-of-view. I manually tested that it works on the table, but there isn't an auto test for it.
- TanStack does not accept an empty string for the ID, so I don't see a good path forward of accepting this as a valid ID on the `nimble-table`. I have validation code & a corresponding test for this. I also included this in the docs.
- TanStack only allows strings for the row IDs. I'd prefer not to do conversion between the `nimble-table` IDs and TanStack IDs, so I think we should stick with requiring string IDs unless there is a strong reason that we need support another data type.